flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/RowDataToStringConverterImpl.java

```diff
@@ -51,9 +53,18 @@ public RowDataToStringConverterImpl(DataType dataType) {
                 false);
     }
 
-    @SuppressWarnings("unchecked")
     public RowDataToStringConverterImpl(
             DataType dataType, ZoneId zoneId, ClassLoader classLoader, boolean legacyBehaviour) {
+        this.dataType = dataType;
+        this.castRuleContext = CastRule.Context.create(legacyBehaviour, zoneId, classLoader);
+    }
+
+    @SuppressWarnings("unchecked")
+    public void init() {
+        if (this.columnConverters != null) {
```

Comment:
if we move this condition into `convert`, we safe a virtual call in the hot path